### ISSUE TITLE
Is it possible to read from one topic and write to another within a single Spark Structured Streaming job?

### ISSUE THREAD

**Issue by brandon-stanley** [AbsaOSS/ABRiS#169]
Hi ABRiS Team,

I am trying to read from one Confluent Kafka topic (registered with Schema Registry) and write to another Confluent Kafka topics (also registered with Schema Registry) within my Spark Structured Streaming job. The two topics have different schemas.  I noticed that writing records from the source topic to the target topic are slow (~10 records per second). However, when I run a job that only reads from the source topic or another job that only writes (dummy records) to the target topic, I do not see the performance issues. The issue only arises when I try to read from one topic and write to another. Is this something that is not achievable? 

Cheers,

Brandon

**Comment by cerveada** [comment]
Hello,

- What version of Spark and Abris do you use?
- Could you provide the code that you use for that?

**Comment by brandon-stanley** [comment]
Spark 2.4.0
ABRiS 3.2.1

Here is a snippet of the code I am using:

`
from pyspark.sql import SparkSession
from pyspark.sql import Column, _to_java_column
from pyspark.sql.column import *
from pyspark.sql import functions as F
from pyspark import SparkContext
from pyspark.sql.functions import col


def to_avro(col, conf_map):
    """
    avro  serialize

    :param col: column name "key" or "value"
    :param conf_map: configuration map for schema
    :return:
    """
    jvm_gateway = SparkContext._active_spark_context._gateway.jvm
    abris_avro = jvm_gateway.za.co.absa.abris.avro
    return Column(
        abris_avro.functions.to_confluent_avro(_to_java_column(col), conf_map)
    )


def from_avro(col, conf_map):
    """
    avro deserialize

    :param col: column name "key" or "value"
    :param conf_map: configuration map for schema
    :return:
    """
    jvm_gateway = SparkContext._active_spark_context._gateway.jvm
    abris_avro = jvm_gateway.za.co.absa.abris.avro
    return Column(
        abris_avro.functions.from_confluent_avro(_to_java_column(col), conf_map)
    )


def confluent_avro_deserialize(sc, df, conf_map):
    deserialized_df = df.select(
        col("key").cast("string").alias("_kafka_key"),
        col("partition").alias("_kafka_partition"),
        col("offset").alias("_kafka_offset"),
        col("timestamp").alias("_kafka_timestamp"),
        col("timestampType").alias("_kafka_timestampType"),
        from_avro(df.value, conf_map).alias("value"),
    )
    return deserialized_df


class SchemaRegistry:
    def __init__(
        self,
        spark,
        schema_registry_url,
        schema_registry_topic,
        value_schema_naming_strategy="topic.name",
        value_schema_id="latest",
    ):
        self.spark = spark
        self.sc = spark.sparkContext
        self.schema_registry_url = schema_registry_url
        self.schema_registry_topic = schema_registry_topic
        self.value_schema_naming_strategy = value_schema_naming_strategy
        self.value_schema_id = value_schema_id

    def get_conf_map(self, write=False):
        jvm_gateway = self.sc._gateway.jvm
        abris_avro = jvm_gateway.za.co.absa.abris.avro
        naming_strategy = getattr(
            getattr(
                abris_avro.read.confluent.SchemaManager,
                "SchemaStorageNamingStrategies$",
            ),
            "MODULE$",
        ).TOPIC_NAME()

        if write:
            schema_registry_config_dict = {
                "schema.registry.url": self.schema_registry_url,
                "schema.registry.topic": self.schema_registry_topic,
                "value.schema.naming.strategy": naming_strategy,
            }
        else:
            schema_registry_config_dict = {
                "schema.registry.url": self.schema_registry_url,
                "schema.registry.topic": self.schema_registry_topic,
                "value.schema.id": self.value_schema_id,
                "value.schema.naming.strategy": self.value_schema_naming_strategy,
            }

        conf_map = getattr(
            getattr(jvm_gateway.scala.collection.immutable.Map, "EmptyMap$"), "MODULE$"
        )

        for k, v in schema_registry_config_dict.items():
            conf_map = getattr(conf_map, "$plus")(jvm_gateway.scala.Tuple2(k, v))

        return conf_map


spark = SparkSession.builder.appName("appName").enableHiveSupport().getOrCreate()
sc = spark.sparkContext

src_topic = "<SRC_TOPIC>"
schema_registry_url = "<SCHEMA_REGISTRY_URL>"
src_read_options = {"startingOffsets": "earliest"}
src_df = spark.readStream.option("subscribe", src_topic).load(
    format="kafka", **src_read_options
)

# Create schema registry object
schema_registry_src = SchemaRegistry(
    spark=spark,
    schema_registry_url=schema_registry_url,
    schema_registry_topic=src_topic,
)
conf_map_src = schema_registry_src.get_conf_map()

deserialized_df_src = confluent_avro_deserialize(
    sc=sc, df=src_df, conf_map=conf_map_src
)

kafka_key_list = ["<KEY_COL_1>", "KEY_COL_2"]

deserialized_df_src = deserialized_df_src.select("value.*")
deserialized_df_src = deserialized_df_src.withColumn("key", F.concat(*kafka_key_list))


reserialized_df = deserialized_df_src.select(
    "key", to_avro(deserialized_df_src.value, conf_map_src).alias("value")
)

dest_topic = "<DEST_TOPIC>"
schema_registry_dest = SchemaRegistry(
    spark=spark,
    schema_registry_url=schema_registry_url,
    schema_registry_topic=dest_topic,
)

conf_map_dest = schema_registry_dest.get_conf_map(write=True)

write_options = {"checkpointLocation": "<CHECKPOINT_PATH>", "topic": dest_topic}
write = reserialized_df.selectExpr("key", "value").writeStream
write = write.trigger(once=True)
write = write.start(format="kafka", **write_options)

write.awaitTermination()
`

**Comment by cerveada** [comment]
Shouldn't `to_avro` use `conf_map_dest` instead of `conf_map_src`?

I don't know for sure what is causing the performance issues, but I suspect it's some communication between Abris and schema registry. 

You can try Abris version  3.2.2 where some of it was optimized.

Abris version 4 is even better in this and should be most efficient, but currently there are some conflicting libraries with Spark, but this will be fixed in 4.0.1 that is coming soon. Version 4 is using a different API so you will have to change the code a bit to use it.

**Comment by brandon-stanley** [comment]
Thanks for pointing this out. `to_avro` is using `conf_map_dest`. I copied + pasted incorrectly. 

I think the performance issue has something to do with "switching" between the two `schemaRegistryConfig` object instances for the src/dest topics. Perhaps, the JVM object (for SchemaRegistry) is shared, and each batch of records is forcing an API call to SchemaRegistry (to apply the correct schema for the src and target records).

I will try `3.2.2` to see if there are any differences.

**Comment by brandon-stanley** [comment]
Unfortunately, there was no change with `3.2.2`

**Comment by brandon-stanley** [comment]
@felipemmelo I see that you've commented about a similar issue in this ticket: https://github.com/AbsaOSS/ABRiS/issues/35. Would you be able to provide any insight?

**Comment by cerveada** [comment]
> Also, I know that the APIs have changed for 4.0 but there is no sample code for PySpark (as mentioned in the documentation). Has anyone provided an update for this?

Not yet, but it shouldn't be hard. The main thing that changed is the configuration, that is now created by a chain of function calls instead of a map. Once you have the configuration object the rest is very similar.

The `jvm_gateway` object is from this library https://www.py4j.org/ in case you want to try it.

> I think the performance issue has something to do with "switching" between the two `schemaRegistryConfig` object instances for the src/dest topics. Perhaps, the JVM object (for SchemaRegistry) is shared, and each batch of records is forcing an API call to SchemaRegistry (to apply the correct schema for the src and target records).

This is the kind of issue that the version 4 is supposed to solve.

**Comment by brandon-stanley** [comment]
We were able to resolve the performance issue by adding resource-specific configuration parameters passed into our `spark-submit` command as well as adding `df.coalesce()` after deserializing the data from the source topic:
```
spark-submit ...
--driver-memory 2G \
--executor-memory 2G \
--executor-cores 3 \
--num-executors 3 \
...
```

```
deserialized_df_src = deserialized_df_src.select("value.*")
deserialized_df_src = deserialized_df_src.coalesce(3)
```
The records' size was most likely the root cause of this is the size of the record. The records have 400+ columns, and there may have been shuffling/issues memory allocation issues. After altering the configurations and code, as mentioned above, we saw a major increase in performance.
